test(eap): Testing framework for running mutations on Clickhouse

ayirr7 opened the pull request:
A follow-up to #6216 

More rigorous testing for the main query run on the mutations consumer. We need to be able to catch any correctness issues with the query in CI itself.

untitaker commented on rust_snuba/src/mutations/clickhouse.rs:
```diff
@@ -170,6 +170,76 @@ fn format_query(table: &str, batch: &MutationBatch) -> Vec<Vec<u8>> {
+
+impl ClickhouseTestClient {
+    pub async fn new(table: String) -> anyhow::Result<Self> {
+        // hardcoding local Clickhouse settings
```
i think we want to make those environment variables with these defaults. however, i think the 9000 port is also wrong, you want 8123

ayirr7 replied:
oh yeah i was testing this out and forgot to revert. 8123 is correct

untitaker commented on rust_snuba/src/mutations/clickhouse.rs:
```diff
@@ -170,6 +170,76 @@ fn format_query(table: &str, batch: &MutationBatch) -> Vec<Vec<u8>> {
+        let url = format!("http://{hostname}:{http_port}");
+
+        let client = reqwest::Client::new();
+        let test_table = format!("{table}_test_local\n");
```
this should contain a UUID or otherwise random value. we will have many tests that use this client, and `cargo test` runs them in parallel. if the tests were run sequentially, it would not be necessary to create a separate table per test at all

ayirr7 replied:
Oh, I didn't realize that `cargo test` would run them in parallel. Thanks for calling that out

ayirr7 replied:
Done

untitaker commented on rust_snuba/src/mutations/clickhouse.rs:
```diff
@@ -170,6 +170,76 @@ fn format_query(table: &str, batch: &MutationBatch) -> Vec<Vec<u8>> {
+        Ok(())
+    }
+
+    pub async fn drop_table(&self) -> anyhow::Result<()> {
```
do `impl Drop for ClickhouseTestClient`, and inside `fn drop()`, run `tokio::block_on(self.drop_table())`. then this function does not need to be called explicitly from the test, the table will just disappear when the test is over

ayirr7 replied:
I tried this and for some reason it's causing the test to time out (hang) altogether, i.e:

`test mutations::clickhouse::tests::test_simple_mutation has been running for over 60 seconds`

I also tried a different variation and ran into `Cannot start a runtime from within a runtime.` I can dig deeper into this to understand what would work, but it seems a manual drop could also do? Let me know if you find another approach that works -- for supporting an asynchronous drop function

untitaker replied:
Ok let's do it manually for now, we can fix it later

ayirr7 replied:
I agree, I'd like to fix it so that I don't have to manually call drop. I'll keep looking into this on the side and if I find a solution that works I'll fix it.

untitaker commented on rust_snuba/src/mutations/clickhouse.rs:
```diff
@@ -170,6 +170,76 @@ fn format_query(table: &str, batch: &MutationBatch) -> Vec<Vec<u8>> {
+        Ok(())
+    }
+
+    pub async fn insert_data(&self, primary_key: PrimaryKey) -> anyhow::Result<()> {
```
I don't think this abstraction is necessary, I believe it's easier (or at least more flexible) to write the entire query yourself in the test, and invoke `run_query`

ayirr7 replied:
Removed

untitaker commented on rust_snuba/src/mutations/clickhouse.rs:
```diff
@@ -170,6 +170,76 @@ fn format_query(table: &str, batch: &MutationBatch) -> Vec<Vec<u8>> {
+    pub async fn run_query(&self, queries: Vec<Vec<u8>>) -> anyhow::Result<()> {
+        // run the vec of queries against Clickhouse
+        for q in queries {
+            self.client
```
this needs to have session_id as an attribute, otherwise the temporary table will disappear

ayirr7 commented:
I've addressed the comments locally. However I think I'm still running into some issues with the core functionality of the test (I described them offline)

untitaker commented:
1. moved ClickhouseTestClient into the `mod tests{}`, then you don't have to use `allow(dead_code)` as it is only compiled when the tests are being compiled (`cfg(test)`)
2. removed some comments
3. fixed the CI issue -- in CI, the clickhouse is available on `clickhouse`, not `localhost`: https://github.com/getsentry/snuba/blob/bc31d9544380581d77309087a0e000b2e60408d4/docker-compose.gcb.yml#L20

so i added a check for that environment variable

the mutations change looks good. i think the PR is ready to merge
